test/test_common.c

```diff
@@ -5,3 +5,148 @@
  */
 
 #include "test_common.h"
+
+void *rpmi_env_zalloc(rpmi_size_t size)
+{
+	return malloc(size);
+}
+
+void rpmi_env_free(void *ptr)
+{
+	free(ptr);
+}
+
+void test_run(struct rpmi_test *rpmi_test)
+{
+	struct rpmi_message *req_msg = get_rpmi_msg(rpmi_test->rpmi_xport);
+
+	req_msg->header.servicegroup_id = rpmi_test->attrs.svc_grp_id;
+	req_msg->header.service_id = rpmi_test->attrs.svc_id;
+	req_msg->header.datalen = rpmi_test->query_data_len;
+	rpmi_env_memcpy(req_msg->data, rpmi_test->query_data, rpmi_test->query_data_len);
+
+	test_xport_enqueue(rpmi_test->rpmi_xport, RPMI_QUEUE_A2P_REQ, req_msg);
+
+	free_rpmi_msg(req_msg);
+}
+
+void test_wait(struct rpmi_test *rpmi_test, struct rpmi_message *msg)
+{
+	int result = -1;
+
+	rpmi_env_memset(msg, 0, sizeof(*msg));
+
+	/* wait for response */
+	if (rpmi_test->attrs.svc_type != RPMI_MSG_POSTED_REQUEST) {
+
+		while (result != RPMI_SUCCESS)
+			result = test_xport_dequeue(rpmi_test->rpmi_xport,
+						    RPMI_QUEUE_P2A_ACK,
+						    msg);
+	}
+}
+
+void test_verify(struct rpmi_test *rpmi_test, struct rpmi_message *msg)
+{
+	int failed = 0;
+
+	if (msg->header.datalen != rpmi_test->resp_data_len) {
+		/* expected data error */
+		printf("%s: datalen mismatch: expected: %d, got: %d\n",
+		       rpmi_test->name,
+		       rpmi_test->resp_data_len, msg->header.datalen);
+		failed = 1;
+	} else {
+		if (rpmi_env_memcmp(&rpmi_test->resp_data, &msg->data,
+			   rpmi_test->resp_data_len)) {
+			printf("%s: datalen: %d, expected data mismatch\n",
+			       rpmi_test->name,
+			       rpmi_test->resp_data_len);
+			hexdump("expected",
+				(unsigned int *)&rpmi_test->resp_data,
+				rpmi_test->resp_data_len);
+			hexdump("received",
+				(unsigned int *)msg->data,
+				rpmi_test->resp_data_len);
+			failed = 1;
+		}
+	}
+	printf("%-50s \t : %s!\n", rpmi_test->name, failed?"Failed":"Succeeded");
+}
+
+void execute_rpmi_test_group(struct rpmi_transport *rpmi_xport, struct rpmi_test_group *group)
```

Comment:
We can improve this common test infrastructure such that each test has both client and server part of the test-case. This will help us avoid creating a separate server application for each test scenario.